jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java

```diff
@@ -123,52 +131,49 @@ public Node getNode() {
 				this.nodeInstanceContainer.getNodeContainer()).internalGetNode( this.nodeId );
     	} catch (IllegalArgumentException e) {
     		throw new IllegalArgumentException(
-				"Unknown node id: " + this.nodeId 
+				"Unknown node id: " + this.nodeId
 				+ " for node instance " + getUniqueId()
 				+ " for process instance " + this.processInstance, e);
     	}
     }
-    
+
     public boolean isInversionOfControl() {
         return false;
     }
-    
+
     public void cancel() {
         nodeInstanceContainer.removeNodeInstance(this);
-        boolean hidden = false;
-        Node node = getNode();
-    	if (node != null && node.getMetaData().get("hidden") != null) {
-    		hidden = true;
-    	}
+        if( isStackless() ) {
```

Comment:
could we move this isStackless chekgin into the removeNodeInstanceTrigger method instead to have less cleanup work afterwards?